internal/app/fipcontroller/kubernetes.go

```diff
@@ -31,32 +31,79 @@ func newKubernetesClient() (*kubernetes.Clientset, error) {
  * Search and return the IP address of a given kubernetes node name.
  *  Will return first found internal or external IP depending on nodeAddressType parameter
  */
-func (controller *Controller) nodeAddress(ctx context.Context, nodeName string, nodeAddressType configuration.NodeAddressType) (address net.IP, err error) {
-	nodes, err := controller.KubernetesClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+func (controller *Controller) nodeAddressList(ctx context.Context, nodeAddressType configuration.NodeAddressType) (addressList []net.IP, err error) {
+	// Try to get deployment pods if certain label is specified
+	listOptions := metav1.ListOptions{}
+	listOptions.LabelSelector = controller.Configuration.PodLabelSelector
+
+	pods, err := controller.KubernetesClient.CoreV1().Pods(controller.Configuration.Namespace).List(ctx, listOptions)
+	if err != nil {
+		return nil, fmt.Errorf("could not list nodes: %v", err)
+	}
+	controller.Logger.Debugf("Found %d pods", len(pods.Items))
+
+	for _, pod := range pods.Items {
+		address := net.ParseIP(pod.Status.HostIP)
+		addressList = append(addressList, address)
+	}
+	if len(addressList) > 0 {
+		controller.Logger.Debugf("Found %d ips from pods", len(addressList))
+		return
+	}
+
+	// Create list options with optional labelSelector
+	listOptions = metav1.ListOptions{}
+	if controller.Configuration.NodeLabelSelector != "" {
+		listOptions.LabelSelector = controller.Configuration.NodeLabelSelector
+	}
+
+	nodes, err := controller.KubernetesClient.CoreV1().Nodes().List(ctx, listOptions)
 	if err != nil {
 		return nil, fmt.Errorf("could not list nodes: %v", err)
 	}
 	controller.Logger.Debugf("Found %d nodes", len(nodes.Items))
 
-	var addresses []corev1.NodeAddress
 	for _, node := range nodes.Items {
-		if node.Name == nodeName {
-			addresses = node.Status.Addresses
-			break
+		// Skip unhealthy nodes
+		if !isNodeHealthy(node) {
```

Comment:
Nitpick: This might result in an empty addressList (when the nodes are running again, but not yet reporting healthy to the API Server)

Reply:
the method will now return an error, if no healthy nodes can be found